### PR TITLE
Add liquidate vault

### PR DESCRIPTION
**Pull request by alannotnerd**
* add liquidate vault  
Liquidating vault on block initialize.

### CODE REVIEW

File: xpallets/gateway/bitcoin/v2/src/vault.rs [hunk 1]
```diff
@@ -292,5 +293,33 @@ pub mod pallet {
                 Err(Error::<T>::VaultInactive.into())
             }
         }
+
+        /// Liquidate vault and mark it as `Liquidated`
+        ///
+        /// Liquidated vault cannot be updated.
+        pub(crate) fn liquidate_vault(id: &T::AccountId) -> Result<(), DispatchError> {
```
Comment (liuchengxu) [inline xpallets/gateway/bitcoin/v2/src/vault.rs:312]: https://github.com/chainx-org/ChainX/blob/e6a28871a59bf71c61fa2daeb35fa2e3eeade387/xpallets/gateway/bitcoin/v2/src/assets.rs#L138

This TODO can be partially fixed now.

File: xpallets/gateway/bitcoin/v2/src/vault.rs [hunk 2]
```diff
@@ -87,7 +88,16 @@ pub mod pallet {
     pub struct Pallet<T>(PhantomData<T>);
 
     #[pallet::hooks]
-    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_finalize(_: BlockNumberFor<T>) {
+            if assets::Pallet::<T>::is_bridge_running() {
+                <Vaults<T>>::translate(|_, vault| {
```
Comment (liuchengxu) [inline xpallets/gateway/bitcoin/v2/src/vault.rs]: Why `translate`?

Comment (alannotnerd) [reply xpallets/gateway/bitcoin/v2/src/vault.rs]: Cause may mutate vault status

Comment (liuchengxu) [reply xpallets/gateway/bitcoin/v2/src/vault.rs]: `translate` is mostly used when doing the migration, if you want to mutate a storage item, just use `mutate`.

File: xpallets/gateway/bitcoin/v2/src/vault.rs [hunk 3]
```diff
@@ -87,7 +90,21 @@ pub mod pallet {
...
+            if assets::Pallet::<T>::is_bridge_running() {
+                let mut liquidate_vault_ids = BTreeSet::new();
+                for (id, vault) in <Vaults<T>>::iter() {
+                    if Self::_check_vault_liquidated(&vault) {
+                        liquidate_vault_ids.insert(id.clone());
+                    }
+                }
+                for id in liquidate_vault_ids {
+                    let _ = Self::liquidate_vault(&id);
+                }
```
Comment (liuchengxu) [inline xpallets/gateway/bitcoin/v2/src/vault.rs]:
```suggestion
                for (id, vault) in <Vaults<T>>::iter() {
                    if Self::collateral_is_under_threshold(&vault) {
                        let _ = Self::liquidate_vault(&id);
                    }
                }
```

File: xpallets/gateway/bitcoin/v2/src/vault.rs [hunk 4]
```diff
@@ -292,5 +309,47 @@ pub mod pallet {
...
+                return false;
+            }
+            let collateral = <assets::CurrencyOf<T>>::reserved_balance(&vault.id);
+            let collateral_ratio =
+                <assets::Pallet<T>>::calculate_collateral_ratio(vault.issued_tokens, collateral);
+            let liquidation_threshold = Self::liquidation_threshold();
+            match collateral_ratio {
+                Ok(collateral_ratio) => collateral_ratio < liquidation_threshold,
+                _ => false,
+            }
```
Comment (liuchengxu) [inline xpallets/gateway/bitcoin/v2/src/vault.rs]:
```suggestion
            <assets::Pallet<T>>::calculate_collateral_ratio(vault.issued_tokens, collateral)
                .map(|collateral_ratio| collateral_ratio < Self::liquidation_threshold())
                .unwrap_or(false)
```

**Pull request merged**